Blocks Toolkit

SusanRatiLane opened the pull request:
Feature adapted from Lynda Tang's graduate work.

Adds project-level property "BlocksToolkit" (internally also called SubsetJSON), which is a JSON string describing UI components and code blocks to be displayed in the Designer and Blocks Editor. This allows projects to limit the options in the GUI to fit the curriculum and not overwhelm students.

jisqyv commented:
When I use the "Export to File" option of the BlocksToolkit it creates a file with the ".txt" extension. Should it use ".json" instead?

jisqyv commented on appinventor/blocklyeditor/src/drawer.js:
```diff
@@ -81,6 +87,209 @@ Blockly.Drawer.buildTree_ = function() {
+    if (window.parent.BlocklyPanel_getComponentInstancePropertyValue) {
+      subsetJsonString = window.parent.BlocklyPanel_getComponentInstancePropertyValue(formName, screenName, "BlocksToolkit");
+    } else {
+      subsetJsonString = JSON.stringify(window.reactGetSubsetString());
```
I don't see where window.reactGetSubsetString is defined?

SusanRatiLane replied:
Huh.  I didn't immediately find it either. I'll poke around at it. This is Lynda's code, but I never had a problem with it.

Console logs should probably be cleaned up a bit too.

jisqyv commented on appinventor/blocklyeditor/src/drawer.js:
```diff
@@ -81,6 +87,209 @@ Blockly.Drawer.buildTree_ = function() {
+    }  
+    console.log("subsetJsonString");
+    console.log(subsetJsonString);
+    var subsetArray = JSON.parse(subsetJsonString);
```
If subsetJsonString is the empty string (which is the case when loading an older project) then this line throws and exception. Should check to see if the string is empty and provide an appropriate define ([] or {}?).

SusanRatiLane replied:
Heh, and I just hit this when testing. Working on it.

PeterMathijssen commented:
In the BlocksToolKit i am missing the extension palet.

SusanRatiLane commented:
> In the BlocksToolKit i am missing the extension palet.

It doesn't support the extensions palette right now. It's worth checking out what that would take. Since extensions can be added and removed, which affects what's in the toolkit, I deemed it out of scope.

However, it's important that extensions palette not be turned off in the Designer, and I don't think i've tested that.

PeterMathijssen commented:
Yes the extension palet is turned of now when using a custom Blocks Toolkit.

SusanRatiLane commented:
> Yes the extension palet is turned of now when using a custom Blocks Toolkit.

OK, this should be fixed, and I believe the fix is on the test server right now.

PeterMathijssen commented:
I now have to extension categories in the palette
![image](https://user-images.githubusercontent.com/32273650/60364294-17a92680-99e6-11e9-935b-26e4c403bbe8.png)

SusanRatiLane commented:
I have fixed some bugs and changed Match Project to match the event/method/property component blocks as well as the components and built-in blocks. This code could use some refinement, but out respect for the upcoming release, I'd like to kick this back into the review queue.
